File: Day-11/LSRK.js
function isRepeatedSubsequence(s, seq, k) {
  let pos = 0;
  for (let i = 0; i < k; i++) {
      for (let c of seq) {
          pos = s.indexOf(c, pos);
          if (pos === -1) return false;
          pos++; // Move past current character
      }
  }
  return true;
}

function longestSubsequenceRepeatedK(s, k) {
  const freq = new Array(26).fill(0);
  for (let c of s) {
      freq[c.charCodeAt(0) - 'a'.charCodeAt(0)]++;
  }

  const availableChars = [];
  for (let i = 0; i < 26; i++) {
      if (freq[i] >= k) {
          availableChars.push(String.fromCharCode('a'.charCodeAt(0) + i));
      }
  }

  let result = "";
  let queue = [""];

  while (queue.length > 0) {
      const curr = queue.shift();

      for (let c of availableChars) {
          const next = curr + c;

          if (isRepeatedSubsequence(s, next, k)) {
              if (
                  next.length > result.length ||
                  (next.length === result.length && next > result)
              ) {
                  result = next;
              }
              queue.push(next);
          }
      }
  }

  return result;
}

// Example usage
console.log(longestSubsequenceRepeatedK("letslets", 2)); // Output: "lets"

File: Day-11/LSRK.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <unordered_map>
#include <algorithm>

using namespace std;

bool isRepeatedSubsequence(const string& s, const string& seq, int k) {
    int pos = 0;
    for (int i = 0; i < k; ++i) {
        for (char c : seq) {
            pos = s.find(c, pos);
            if (pos == string::npos) return false;
            ++pos; // Move past current character
        }
    }
    return true;
}

string longestSubsequenceRepeatedK(string s, int k) {
    vector<int> freq(26, 0);
    for (char c : s) {
        freq[c - 'a']++;
    }

    vector<char> availableChars;
    for (char c = 'a'; c <= 'z'; ++c) {
        if (freq[c - 'a'] >= k) {
            availableChars.push_back(c);
        }
    }

    string result = "";
    queue<string> q;
    q.push("");

    while (!q.empty()) {
        string curr = q.front(); q.pop();

        for (char c : availableChars) {
            string next = curr + c;

            if (isRepeatedSubsequence(s, next, k)) {
                if (next.length() > result.length() ||
                    (next.length() == result.length() && next > result)) {
                    result = next;
                }
                q.push(next);
            }
        }
    }

    return result;
}

int main() {
    cout << longestSubsequenceRepeatedK("letslets", 2) << endl; // Output: lets
    return 0;
}